enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/readreplica/UpstreamDatabaseSelectionStrategy.java

```diff
@@ -17,25 +17,26 @@
  * You should have received a copy of the GNU Affero General Public License
  * along with this program. If not, see <http://www.gnu.org/licenses/>.
  */
-package org.neo4j.causalclustering.messaging.routing;
+package org.neo4j.causalclustering.readreplica;
 
-import org.neo4j.causalclustering.discovery.CoreTopology;
 import org.neo4j.causalclustering.discovery.TopologyService;
 import org.neo4j.causalclustering.identity.MemberId;
+import org.neo4j.helpers.Service;
 
-public class AlwaysChooseFirstMember implements CoreMemberSelectionStrategy
+public abstract class UpstreamDatabaseSelectionStrategy extends Service
 {
-    private final TopologyService discoveryService;
+    protected TopologyService topologyService;
 
-    public AlwaysChooseFirstMember( TopologyService discoveryService)
+    public UpstreamDatabaseSelectionStrategy( String key, String... altKeys )
     {
-        this.discoveryService = discoveryService;
+        super( key, altKeys );
     }
 
-    @Override
-    public MemberId coreMember() throws CoreMemberSelectionException
+    // Service loaded can't inject this via the constructor
+    void setDiscoveryService( TopologyService topologyService )
     {
-        CoreTopology coreTopology = discoveryService.coreServers();
-        return coreTopology.members().iterator().next();
+        this.topologyService = topologyService;
     }
+
+    public abstract MemberId upstreamDatabase() throws UpstreamDatabaseSelectionException;
```

Comment:
Since this is now pipelined perhaps this should now return an Optional rather than throw an exception so that exceptions are not used for control flow?